crypto/Makefile

```diff
@@ -1,11 +1,38 @@
+# CLANG_VERSION is empty if the compiler is not clang-based
+CLANG_VERSION = $(shell $(CC) --version | grep -Po "clang version \K[\d\.]+")
+CLANG_VERSION_MAJOR = $(shell echo $(CLANG_VERSION) | cut -f1 -d.)
+
+# determine specific version ranges
+ifneq ($(CLANG_VERSION),)
+$(if $(shell [ $(CLANG_VERSION_MAJOR) -ge 13 ] && echo "OK"), \
+    $(eval CLANG_AT_LEAST_13 := true), \
+    $(eval CLANG_AT_LEAST_13 := false))
+$(if $(shell [ $(CLANG_VERSION_MAJOR) -ge 14 ] && echo "OK"), \
+    $(eval CLANG_AT_LEAST_14 := true), \
+    $(eval CLANG_AT_LEAST_14 := false))
+endif
+
 ifeq ($(FUZZER),1)
 CC       ?= clang
 LD       ?= $(CC)
 
 SANFLAGS += -fsanitize=fuzzer
-
 CFLAGS   += $(SANFLAGS)
 LDFLAGS  += $(SANFLAGS)
+
+# only clang versions >= 13 support this feature
+ifeq ($(CLANG_AT_LEAST_13),true)
+$(info "info: using -fsanitize-ignorelist")
+CFLAGS   += -fsanitize-ignorelist=fuzzer/sanitizer_ignorelist.txt
+else
+$(info "info: not using -fsanitize-ignorelist")
+endif
+
+# TODO is there a better solution, for example by disabling a specific optimization technique?
+# there is a clang optimization issue in relation with the blake2 code at -fsanitize=undefined
```

Comment:
Can you elaborate on this? Maybe there is a legitimate bug in the code.

Reply:
From what I can tell at the moment, some parts of the `blake2b.c` and `blake2s.c` code are very challenging for the clang compiler optimization under certain sanitizer conditions. The result is either a significant slowdown of compilation or a complete abort, neither of which is desirable during frequent re-compilation, hence the workaround.
I think this is more of a problem on the part of clang than anything done wrong in the code (e.g. undefined behavior).  Ultimately I plan to isolate a good reproducer and report this to LLVM, but since clang is not used to produce production firmwares, I haven't prioritized this topic yet.

In my experience, having `-fsanitize=undefined` without `-fsanitize=address` in combination with a high optimization level is a good way to trigger the worst-case behavior, especially on newer clang versions.

Example:
* `CC=clang-14 FUZZER=1 SANFLAGS="-fsanitize=undefined" OPTFLAGS="-O3 -g0" make fuzzer` 
```
[...]
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'blake2s.c'.
4.	Running pass 'Greedy Register Allocator' on function '@blake2s_compress'
[...]
clang: error: clang frontend command failed with exit code 139 (use -v to see invocation)
```
The compilation failure is with the very new `14.0.0-++20220108071805+3e9919cf5750-1~exp1~20220108071855.114`, but the general effect can also be observed with older versions such as `clang 9.0.1-16.1`, so this is not a recent regression.

Notably, the problem is not libFuzzer-related and can also be triggered on "normal" builds, for example via `CC=clang-14 CFLAGS="-fsanitize=undefined" OPTFLAGS="-O3 -g0"  make`. In my experience, it's less problematic there in practice since adding ASan into the mix makes the problem go away, and on regular builds there are much less reasons to _only_ build with UBSan.

Reply:
Thanks for the explanation. Not what I expected :-). LGTM.

Reply:
Additional debugging shows that the performance problems are also present with the simple `clang-14 -fsanitize=undefined -O1 -o blake2b.o -c blake2b.c` compilation run, so they're not related to any other compiler flags or Trezor-specific settings that we could change.

Adding `__attribute__ ((optnone))` to `blake2b_compress()` makes the problem effectively disappear, compilation times go down from ~30s to less than ~1s for the file at `-O1` for `clang-14`. Gcc `11.2.0` shows no comparable effects when marking the compress function with `__attribute__ ((optimize("0")))`. 

I think this is additional evidence that it is a clang-only problem which does not negatively affect the production binaries.